app/src/main/java/org/schabi/newpipe/player/mediasession/MediaSessionPlayerUi.java

```diff
@@ -0,0 +1,140 @@
+package org.schabi.newpipe.player.mediasession;
+
+import static org.schabi.newpipe.MainActivity.DEBUG;
+
+import android.content.Intent;
+import android.graphics.Bitmap;
+import android.support.v4.media.MediaMetadataCompat;
+import android.support.v4.media.session.MediaSessionCompat;
+import android.util.Log;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.media.session.MediaButtonReceiver;
+
+import com.google.android.exoplayer2.ForwardingPlayer;
+import com.google.android.exoplayer2.ext.mediasession.MediaSessionConnector;
+
+import org.schabi.newpipe.R;
+import org.schabi.newpipe.player.Player;
+import org.schabi.newpipe.player.ui.PlayerUi;
+import org.schabi.newpipe.player.ui.VideoPlayerUi;
+import org.schabi.newpipe.util.StreamTypeUtil;
+
+import java.util.Optional;
+
+public class MediaSessionPlayerUi extends PlayerUi {
+    private static final String TAG = "MediaSessUi";
+
+    private MediaSessionCompat mediaSession;
+    private MediaSessionConnector sessionConnector;
+
+    public MediaSessionPlayerUi(@NonNull final Player player) {
+        super(player);
+    }
+
+    @Override
+    public void initPlayer() {
+        super.initPlayer();
+        destroyPlayer(); // release previously used resources
+
+        mediaSession = new MediaSessionCompat(context, TAG);
+        mediaSession.setActive(true);
+
+        sessionConnector = new MediaSessionConnector(mediaSession);
+        sessionConnector.setQueueNavigator(new PlayQueueNavigator(mediaSession, player));
+        sessionConnector.setPlayer(getForwardingPlayer());
+
+        sessionConnector.setMetadataDeduplicationEnabled(true);
+        sessionConnector.setMediaMetadataProvider(exoPlayer -> buildMediaMetadata());
+    }
+
+    @Override
+    public void destroyPlayer() {
+        super.destroyPlayer();
+        if (sessionConnector != null) {
+            sessionConnector.setPlayer(null);
+            sessionConnector.setQueueNavigator(null);
+            sessionConnector = null;
+        }
+        if (mediaSession != null) {
+            mediaSession.setActive(false);
+            mediaSession.release();
+            mediaSession = null;
+        }
+    }
+
+    @Override
+    public void onBroadcastReceived(final Intent intent) {
+        super.onBroadcastReceived(intent);
+        // TODO decide whether to handle ACTION_HEADSET_PLUG or not
```

Comment:
TODO detected

Reply:
Actually, I don't know. Maybe we should remove this TODO and call it a day, since it has been there since forever and NewPipe's behavior when headsets are plugged or unplugged is currently correct. What do you think?

Reply:
Then I think we should remove it.

If someone has a problem, we can create an issue and handle it there.